### PR TITLE
feat(training): JSONL exporter for ATIF trajectories

### LINKED ISSUE #87: JSONL training-record exporter from archive
Parent: #86 — Open-Weight Code-Review Model (Milestone 1)

## Spec reference

- Spec: [`.planning/specs/open-weight-review-model/SPEC.md`](../SPEC.md)
- Requirement: **R1 — JSONL training-record exporter**
- Constraints in force: **C5** (eval-set exclusion list must be honored at export time), **C8** (repo-licensing discipline — exclude GPL/AGPL repos by default), **C9** (training corpus is the `accepted`-label subset)

## Problem statement

Daydream's archive (`~/.daydream/archive/runs/*`) already contains ATIF v1.6 trajectories, manifests, and (after R3 lands) outcome labels — but there is no path from that archive into a flat, schema-versioned JSONL corpus that a training pipeline can consume. Without an exporter, every downstream piece of M1 (R2 reward filtering, R5 training, R8 bootstrap dataset assembly) is blocked. This is the substrate dependency that gates the rest of the milestone.

## Scope

- A new daydream CLI subcommand that materializes the archive into JSONL training records, written to a user-specified output path.
- A documented, versioned JSONL record schema covering: code context (changed files / diff refs — see [Memory: Don't embed diffs in prompts](../../../../.claude/projects/-Users-ka-github-existential-birds-daydream/memory/feedback_no_diff_in_prompt.md)), review output, fix diff (when present), test outcome, finding-acceptance label, grounding score, and ATIF span boundaries (REASON vs ACT) per `daydream/trajectory.py` ATIF Step shape.
- Filtering at export time by manifest fields: `skill`, stack (from `daydream/deep/detection.py`), `repo_slug`, outcome label, grounding-rate threshold.
- Stratification flag that emits a balanced (or capped) subset across stacks.
- Enforcement of the C5 exclusion list at export time (refuses to emit records from excluded `repo_slug`s) and the C8 licensing guard (refuses GPL/AGPL repos unless `--allow-copyleft <repo>` is passed).
- Schema-version field on every record; the exporter writes a sibling `schema.json` describing the record shape.
- Tests covering: round-trip on a fixture archive, filter/stratify behavior, exclusion-list enforcement, schema-version stamping.

## Out of scope

- Reward computation / quality filtering — that is **R2 (sub-issue #88)**. The exporter writes the raw label + grounding score; the reward module turns them into a composite scalar.
- Auto-labeling logic — that is **R3 (sub-issue #89)**. The exporter consumes whatever `outcome_labels` are already on the manifest.
- The bootstrap data-collection process — that is **R8 (sub-issue #94)**. The exporter is the *output side* of the archive; R8 is the *input side*.
- Training-recipe code (RFT / SFT / KTO) — that is **R5 (sub-issue #91)**.
- The benchmark corpus itself — that is **R4 (sub-issue #90)**. The exporter only reads the exclusion list R4 produces.

## Acceptance criteria

1. `daydream export-jsonl --out <path>` produces a `.jsonl` file where every line is a JSON object that validates against the published `schema.json`.
2. Each record contains, at minimum: `schema_version`, `session_id`, `repo_slug`, `skill`, `stack`, `code_context` (file paths + base SHA, **not** embedded diffs), `review_output`, `fix_diff_ref`, `test_outcome`, `outcome_label`, `grounding_score`, and an ATIF `spans` list with `kind ∈ {REASON, ACT}` boundaries.
3. `--filter` flags (`--skill`, `--repo`, `--label`, `--min-grounding`) reduce the output set predictably, verified by a test that asserts pre/post record counts on a fixture archive.
4. `--stratify-by stack` produces an output where no single stack exceeds the configured cap (default 60% per OQ4 draft), verified by a test.
5. Records whose `repo_slug` is on the C5 exclusion list are **never** emitted, even without explicit filtering — verified by a test that places a benchmark repo in the fixture archive and asserts zero records leak.
6. GPL/AGPL repos are skipped by default; `--allow-copyleft <repo>` opts in explicitly. A test fixture covers both paths.
7. Re-running the exporter against the same archive produces byte-identical output (deterministic ordering — sort by `session_id`).
8. The `schema.json` is checked into the repo at a stable path so consumers can pin a schema version.

## Technical context

- The archive layout is owned by `daydream/archive/manifest.py` (`Manifest` dataclass — see `outcome_labels`, `repo_slug`, `skill`, `grounding_rate`) and the SQLite index. The exporter reads from both, not from the live `.daydream/` runs directory.
- Trajectory span boundaries live in `daydream/trajectory.py` — ATIF `Step` instances already carry distinct REASON content (text/thinking) and ACT content (tool_calls). Per SPEC KD1, "the data shape SAD requires is already present in our recorder output." Surface these spans as-is; do not re-segment.
- Stack detection lives in `daydream/deep/detection.py` (`detect_stacks()`). The exporter should reuse this rather than re-implementing language detection.
- The CLI surface should follow the existing subcommand pattern (`daydream feedback <pr#>`) — argparse subparser registered in `daydream/cli.py`.
- Schema versioning: model after `MANIFEST_SCHEMA_VERSION` in `daydream/archive/manifest.py`. Bump on every breaking change; consumers pin.
- Pass refs, not contents (per [Memory: Don't embed diffs](../../../../.claude/projects/-Users-ka-github-existential-birds-daydream/memory/feedback_no_diff_in_prompt.md)): records carry `(repo_slug, base_sha, head_sha, changed_files[])`, not the diff bytes. The training pipeline materializes diffs from refs when it needs them.

## Dependencies / blockers

- **No upstream blockers** in this set — this is the foundational sub-issue that unblocks R2, R3, R5, and R8.
- Downstream blockees: **#88 (reward module)**, **#89 (auto-labeling)**, **#91 (training)**, **#94 (bootstrap)** all consume this schema. Lock the schema in this PR; later schema bumps require coordination across those PRs.

## Estimated PR size

**M** (300–1000 LOC). New module + CLI subcommand + schema file + tests; touches `cli.py` and reads from `archive/`, `trajectory.py`, `deep/detection.py` but does not modify them.

## Spike or risk callouts

None inherited from the spec. The only design risk is **schema lock-in**: every downstream sub-issue codes against the schema this PR ships. Review the schema explicitly with the planner before merge.

### PR DESCRIPTION
**Pull request by anderskev**
## Summary

- Adds `daydream export-jsonl` — a new CLI subcommand that turns archived ATIF trajectories into a versioned, schema-validated JSONL corpus suitable for training/eval pipelines.
- Introduces the `daydream/training/` package: schema v1 definition, exclusion + copyleft filtering, record/span builders, deterministic stack-stratification, and an end-to-end export orchestrator.
- Extends the archive manifest with a `code_context` block (`base_sha`, `changed_files`) so newly archived runs carry the git context the exporter needs end-to-end.

## Motivation

The trajectory archive captures per-run ATIF data, but there was no supported way to turn it into a training-ready dataset. Downstream consumers need:

- A stable, versioned JSONL schema (v1) they can validate against.
- Deterministic, byte-identical output across reruns so corpora can be diffed and cached.
- Built-in filtering for license-incompatible (copyleft) and explicitly excluded sources.
- Stack-aware stratification so a single dominant stack does not swamp the corpus.

This PR delivers that pipeline in one focused milestone so the training side can consume archived runs without bespoke glue.

### Added
- `daydream/training/` package: `schema.py`, `exclusion.py`, `export.py`, schema artifacts (`schema/v1.json`, `schema/copyleft.txt`, `schema/exclusion.txt`).
- `daydream export-jsonl --out <path>` CLI subcommand with the full filter, stratification, opt-in, and diagnostic flag surface from the plan (validates `--max-stack-share` in `(0, 1]` and `--min-grounding` in `[0, 1]`).
- `ExportConfig` + `run_export` orchestrator: schema-only short-circuit, filter+stratify pipeline, dry-run summary, atomic JSONL write via tempfile + `Path.replace`, and `schema.json` side-car emission.
- `GitContext` + `Manifest` carry `base_sha` and `changed_files`; `git_ops.diff_name_only` populates them at archive time; `manifest.to_dict()` emits a `code_context` block.
- `archive.index.count_runs()` for cheap unfiltered totals (no row materialization).
- Test fixtures under `tests/fixtures/training/` plus suites covering exclusion, query/filter, record/span builders, stratification, and the end-to-end export.

### Changed
- Exporter sources `head_sha` / `branch` / `base_branch` from the manifest `git` block (was inadvertently reading from `code_context`); falls back to `manifest_row` when older archives have no `code_context` block.
- `is_copyleft()` accepts a pre-loaded list so the per-row hot loop in `_query_index` doesn't reopen the file N times.
- `_build_query()` accepts an `exclusion` kwarg (typed `frozenset | set | None`) so callers can inject the set without re-reading from disk.
- Step IDs are cast to `int` to match the v1 schema integer constraint.
- JSONL output uses `sort_keys=True` and compact JSON separators for deterministic, byte-identical reruns.
- `code_context` field order in the manifest now matches the schema definition (`head_sha` then `base_sha`).
- Skip-record warning copy clarified from "missing" to "corrupt or unreadable"; `stack=None` warning is now gated behind `stratify_by == \"stack\"`.

### Fixed
- Atomic tempfile writes are wrapped in try/except so a failed write doesn't leave an orphan `.tmp` file beside the output.
- Multi-outcome sessions go through `_single_outcome_label()`, which warns instead of silently dropping labels.
- Schema v1 drops the unused `test_outcome` field.

## Test Plan

- [x] `uv run pytest` — full suite passes locally (343 existing + new training/archive/git_ops coverage).
- [x] `uv run ruff check` and `uv run mypy daydream` pass; the `_build_query` exclusion-param typing fix was added specifically to unblock mypy.
- [x] Exporter determinism verified via a fixture-driven byte-identical rerun test.
- [x] Schema-only short-circuit, dry-run no-op, and missing-trajectory skip paths each covered by a dedicated test.
- [x] `git_ops.diff_name_only` covered for happy path, multi-file output, empty-line filtering, and bad-ref soft-failure.

## Checklist

- [x] Tests pass locally (`uv run pytest`)
- [x] Linting passes (`uv run ruff check`)
- [x] Type checking passes (`uv run mypy daydream`)
- [x] Documentation updated (if applicable) — CLI surface documented inline; no user-facing docs site to update for this milestone.

## Additional Context

Diff size: ~2,100 LOC added across 22 files; the bulk lives in `daydream/training/export.py` and its test suite. The `code_context` rollout is forward-compatible: older archives that predate the block surface `base_sha=None` / `changed_files=[]`, which the v1 schema explicitly allows.

---

Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

**Review by anderskev — COMMENTED** [review]
**Code Review Summary**

<details>
<summary>📋 Non-inline findings (11)</summary><blockquote>

<details>
<summary>daydream/archive/index.py (1)</summary><blockquote>

**`changed_files` missing from SQLite schema makes `FixAppliedSignal` permanently inert**

The `runs` table schema has no `changed_files` column, and `_UPSERT_SQL` does not insert it. The `Manifest` dataclass carries `changed_files` in memory and writes it to `manifest.json` on disk, but this data never reaches SQLite. Consequently, `_row_changed_files()` in `daydream/training/labeler.py` (line 325) always returns `[]`, `_safe_fix_applied()` short-circuits (line 251), and the entire fix-applied cascade — hunk-matching logic, 50% threshold, `window_days` parameter — is dead code in production. The labeler test passes because the outcome label is determined by `pr_merge.merged=True` + `comment_resolution.unresolved=0`, making the fix-applied signal irrelevant to the assertion. Also affects `daydream/training/labeler.py`. Either (a) add `changed_files TEXT` (JSON-encoded list) and `base_sha TEXT` columns to the `runs` table schema and `_UPSERT_SQL`, persisting them from the `Manifest` at upsert time, or (b) have the labeler read `manifest.json` from `row["archive_path"]` to retrieve `code_context.changed_files` at labeling time. Option (a) is preferred. Add a test that asserts `fix_applied.verdict == "applied"` to prevent this regression class. (Sources: python-records item 1, alternatives item 1)

<details>
<summary>🔮 Prompt for AI Agents</summary>

```
Verify each finding against the current code and only fix it if needed.

In `daydream/archive/index.py`, `changed_files` missing from SQLite schema makes `FixAppliedSignal` permanently inert: The `runs` table schema has no `changed_files` column, and `_UPSERT_SQL` does not insert it. The `Manifest` dataclass carries `changed_files` in memory and writes it to `manifest.json` on disk, but this data never reaches SQLite. Consequently, `_row_changed_files()` in `daydream/training/labeler.py` (line 325) always returns `[]`, `_safe_fix_applied()` short-circuits (line 251), and the entire fix-applied cascade — hunk-matching logic, 50% threshold, `window_days` parameter — is dead code in production. The labeler test passes because the outcome label is determined by `pr_merge.merged=True` + `comment_resolution.unresolved=0`, making the fix-applied signal irrelevant to the assertion. Also affects `daydream/training/labeler.py`. Either (a) add `changed_files TEXT` (JSON-encoded list) and `base_sha TEXT` columns to the `runs` table schema and `_UPSERT_SQL`, persisting them from the `Manifest` at upsert time, or (b) have the labeler read `manifest.json` from `row["archive_path"]` to retrieve `code_context.changed_files` at labeling time. Option (a) is preferred. Add a test that asserts `fix_applied.verdict == "applied"` to prevent this regression class. (Sources: python-records item 1, alternatives item 1)
```

</details>

</blockquote></details>
<details>
<summary>daydream/training/labeler.py (3)</summary><blockquote>

**`_local_branch_verdict` returns `rejected` when no commits exist, but docstring says `unknown`**

The docstring (line 219) states `"unknown" when no commits to inspect`, but the implementation (lines 221–222) returns `LocalCommitAppliedSignal(verdict="rejected")`. This causes a PR-less run with no subsequent commits to be permanently labeled `"rejected"` instead of `"unknown"`. Downstream in `rubric.py` (lines 113–114), `verdict="rejected"` maps to outcome label `"rejected"`, while `"unknown"` would map to `"unknown"`. Align implementation with the docstring: return `verdict="unknown"` when `not commits`, and update `test_local_commit_applied_signal_no_local_commits_returns_rejected` to expect `"unknown"`. (Sources: python-records item 2, alternatives item 2)

<details>
<summary>🔮 Prompt for AI Agents</summary>

```
Verify each finding against the current code and only fix it if needed.

In `daydream/training/labeler.py` around line 221, `_local_branch_verdict` returns `rejected` when no commits exist, but docstring says `unknown`: The docstring (line 219) states `"unknown" when no commits to inspect`, but the implementation (lines 221–222) returns `LocalCommitAppliedSignal(verdict="rejected")`. This causes a PR-less run with no subsequent commits to be permanently labeled `"rejected"` instead of `"unknown"`. Downstream in `rubric.py` (lines 113–114), `verdict="rejected"` maps to outcome label `"rejected"`, while `"unknown"` would map to `"unknown"`. Align implementation with the docstring: return `verdict="unknown"` when `not commits`, and update `test_local_commit_applied_signal_no_local_commits_returns_rejected` to expect `"unknown"`. (Sources: python-records item 2, alternatives item 2)
```

</details>

---

**Labeler accesses private `git_ops._run_git` instead of public API**

Both `_commits_in_window()` (line 83) and `_commits_since()` (line 106) call `git_ops._run_git()` — a private function — with `# noqa: SLF001` suppression. The existing public `git_ops.log()` only supports `base..head --oneline` format. Add thin public wrappers in `git_ops.py` (e.g. `log_shas(repo, *, since, ref)`) that encapsulate the `git log --pretty=%H` patterns the labeler needs, removing the private-API coupling. (Source: alternatives item 4)

<details>
<summary>🔮 Prompt for AI Agents</summary>

```
Verify each finding against the current code and only fix it if needed.

In `daydream/training/labeler.py` around line 83, Labeler accesses private `git_ops._run_git` instead of public API: Both `_commits_in_window()` (line 83) and `_commits_since()` (line 106) call `git_ops._run_git()` — a private function — with `# noqa: SLF001` suppression. The existing public `git_ops.log()` only supports `base..head --oneline` format. Add thin public wrappers in `git_ops.py` (e.g. `log_shas(repo, *, since, ref)`) that encapsulate the `git log --pretty=%H` patterns the labeler needs, removing the private-API coupling. (Source: alternatives item 4)
```

</details>

---

**`_gh_api` ignores `repo` parameter, preventing multi-host GitHub support**

`_gh_api(repo, endpoint, **kwargs)` accepts a `repo` slug but discards it, always calling `git_ops.gh_api(Path("."), endpoint, **kwargs)`. This works for single-host GitHub.com usage but would break for GitHub Enterprise instances requiring `--hostname`. Acceptable for R1 scope — add a comment documenting the limitation and the path to multi-host support. (Source: alternatives item 8)

<details>
<summary>🔮 Prompt for AI Agents</summary>

```
Verify each finding against the current code and only fix it if needed.

In `daydream/training/labeler.py` around line 64, `_gh_api` ignores `repo` parameter, preventing multi-host GitHub support: `_gh_api(repo, endpoint, **kwargs)` accepts a `repo` slug but discards it, always calling `git_ops.gh_api(Path("."), endpoint, **kwargs)`. This works for single-host GitHub.com usage but would break for GitHub Enterprise instances requiring `--hostname`. Acceptable for R1 scope — add a comment documenting the limitation and the path to multi-host support. (Source: alternatives item 8)
```

</details>

</blockquote></details>
<details>
<summary>daydream/training/snapshot.py (2)</summary><blockquote>

**Snapshot `archive_index_sha` computation ignores WAL journal data**

`run_snapshot()` computes `hashlib.sha256(fp.read())` over `index.db`, but the archive uses WAL mode (`PRAGMA journal_mode=WAL` at `index.py` line 147). Pending or recently committed data may reside in `index.db-wal` rather than the main file. The SHA-256 will not reflect the true database state until WAL checkpointing occurs, and the same logical state produces different file hashes before and after checkpoint — breaking the reproducibility guarantee the snapshot is designed to provide. Before computing the hash, force a WAL checkpoint via `PRAGMA wal_checkpoint(TRUNCATE)`, then read the file. (Sources: python-records item 3, alternatives item 3)

<details>
<summary>🔮 Prompt for AI Agents</summary>

```
Verify each finding against the current code and only fix it if needed.

In `daydream/training/snapshot.py` around line 82, Snapshot `archive_index_sha` computation ignores WAL journal data: `run_snapshot()` computes `hashlib.sha256(fp.read())` over `index.db`, but the archive uses WAL mode (`PRAGMA journal_mode=WAL` at `index.py` line 147). Pending or recently committed data may reside in `index.db-wal` rather than the main file. The SHA-256 will not reflect the true database state until WAL checkpointing occurs, and the same logical state produces different file hashes before and after checkpoint — breaking the reproducibility guarantee the snapshot is designed to provide. Before computing the hash, force a WAL checkpoint via `PRAGMA wal_checkpoint(TRUNCATE)`, then read the file. (Sources: python-records item 3, alternatives item 3)
```

</details>

---

**Snapshot issues N+1 queries for label observations**

`run_snapshot()` first calls `query_runs()` to load all rows, then loops calling `latest_label_observation()` per row — one SQLite query per run. For archives with many runs this scales poorly and is unnecessary; a single aggregate query (`SELECT ... GROUP BY session_id HAVING observed_at = MAX(observed_at)`) could compute label counts directly. Add a `label_count_summary(archive_dir, as_of)` function to `index.py` that returns label counts in one query. (Source: alternatives item 5)

<details>
<summary>🔮 Prompt for AI Agents</summary>

```
Verify each finding against the current code and only fix it if needed.

In `daydream/training/snapshot.py` around line 87, Snapshot issues N+1 queries for label observations: `run_snapshot()` first calls `query_runs()` to load all rows, then loops calling `latest_label_observation()` per row — one SQLite query per run. For archives with many runs this scales poorly and is unnecessary; a single aggregate query (`SELECT ... GROUP BY session_id HAVING observed_at = MAX(observed_at)`) could compute label counts directly. Add a `label_count_summary(archive_dir, as_of)` function to `index.py` that returns label counts in one query. (Source: alternatives item 5)
```

</details>

</blockquote></details>
<details>
<summary>daydream/training/rubric.py (1)</summary><blockquote>

**`derive_outcome_label` uses `assert` for runtime invariant that can be violated**

The function uses `assert local_commit_applied is not None` to guard the local-branch code path. Under Python's `-O` flag, `assert` statements are stripped. If `local_commit_applied` is `None` but `posterior_source="local_branch"`, the code would proceed to `local_commit_applied.verdict`, raising an `AttributeError` instead of a controlled failure. Replace `assert` with an explicit `if ... is None: raise ValueError(...)` or a descriptive `RuntimeError`. (Source: python-records item 4)

<details>
<summary>🔮 Prompt for AI Agents</summary>

```
Verify each finding against the current code and only fix it if needed.

In `daydream/training/rubric.py` around line 111, `derive_outcome_label` uses `assert` for runtime invariant that can be violated: The function uses `assert local_commit_applied is not None` to guard the local-branch code path. Under Python's `-O` flag, `assert` statements are stripped. If `local_commit_applied` is `None` but `posterior_source="local_branch"`, the code would proceed to `local_commit_applied.verdict`, raising an `AttributeError` instead of a controlled failure. Replace `assert` with an explicit `if ... is None: raise ValueError(...)` or a descriptive `RuntimeError`. (Source: python-records item 4)
```

</details>

</blockquote></details>
<details>
<summary>tests/test_training_labeler.py (1)</summary><blockquote>

**Redundant and inconsistent inline `import json` statements in test functions**

The test file contains multiple inline `import json` statements within test function bodies (lines 49, 79) plus a duplicate `import json as _json` (line 83), violating the project's top-of-file import organization convention (stdlib → third-party → local). Move `import json` to the top-level imports and remove the inline duplicates. (Sources: python-records item 5, alternatives item 6)

<details>
<summary>🔮 Prompt for AI Agents</summary>

```
Verify each finding against the current code and only fix it if needed.

In `tests/test_training_labeler.py`, Redundant and inconsistent inline `import json` statements in test functions: The test file contains multiple inline `import json` statements within test function bodies (lines 49, 79) plus a duplicate `import json as _json` (line 83), violating the project's top-of-file import organization convention (stdlib → third-party → local). Move `import json` to the top-level imports and remove the inline duplicates. (Sources: python-records item 5, alternatives item 6)
```

</details>

</blockquote></details>
<details>
<summary>tests/contract/fixtures/canonical_script.json (1)</summary><blockquote>

**Contract test canonical script lacks structural variety for the cases it gates**

The canonical script defines only two turns: one single-tool-call turn and one text-only turn. It does not exercise: (a) multiple tool calls in a single turn, (b) tool results arriving after `TurnEndEvent` (the cross-boundary case `_amend_closed_step_observation` handles), (c) error tool results (`is_error: true`), or (d) reasoning-only turns. The cross-turn-boundary case is the critical invariant the `_in_flight_tools` redesign was built for — and it's the one structural case the "parallel-implementation gate" doesn't cover. Also affects `tests/contract/test_backend_step_parity.py`. Add at least one fixture exercising the cross-turn-boundary tool result pattern. (Sources: python-records item 6, alternatives item 7)

<details>
<summary>🔮 Prompt for AI Agents</summary>

```
Verify each finding against the current code and only fix it if needed.

In `tests/contract/fixtures/canonical_script.json`, Contract test canonical script lacks structural variety for the cases it gates: The canonical script defines only two turns: one single-tool-call turn and one text-only turn. It does not exercise: (a) multiple tool calls in a single turn, (b) tool results arriving after `TurnEndEvent` (the cross-boundary case `_amend_closed_step_observation` handles), (c) error tool results (`is_error: true`), or (d) reasoning-only turns. The cross-turn-boundary case is the critical invariant the `_in_flight_tools` redesign was built for — and it's the one structural case the "parallel-implementation gate" doesn't cover. Also affects `tests/contract/test_backend_step_parity.py`. Add at least one fixture exercising the cross-turn-boundary tool result pattern. (Sources: python-records item 6, alternatives item 7)
```

</details>

</blockquote></details>
<details>
<summary>daydream/cli.py (1)</summary><blockquote>

**`_handle_label_command` discards labeler summary return value**

The labeler's `run_backfill()` returns a `dict[str, int]` with counts (labeled, skipped, errored), but `_handle_label_command` silently discards it. `--dry-run` produces zero output on success, unlike the export handler which prints a summary. Print the summary dict to stdout so users get feedback on what the labeler did (or would do in dry-run mode). (Source: python-records item 7)

<details>
<summary>🔮 Prompt for AI Agents</summary>

```
Verify each finding against the current code and only fix it if needed.

In `daydream/cli.py` around line 845, `_handle_label_command` discards labeler summary return value: The labeler's `run_backfill()` returns a `dict[str, int]` with counts (labeled, skipped, errored), but `_handle_label_command` silently discards it. `--dry-run` produces zero output on success, unlike the export handler which prints a summary. Print the summary dict to stdout so users get feedback on what the labeler did (or would do in dry-run mode). (Source: python-records item 7)
```

</details>

</blockquote></details>
<details>
<summary>daydream/training/backfill_cache.py (1)</summary><blockquote>

**`BackfillCache` key does not include `kwargs`, risking cache collisions**

The cache key appears to be `(repo, endpoint)` based on test patterns. If two calls share the same endpoint but differ in `kwargs` (e.g., different `method` or `paginate` parameters), the cache returns stale data. Not triggered in current usage since the labeler always calls endpoints with consistent kwargs, but it's a correctness hazard for future callers. Include `json.dumps(kwargs, sort_keys=True)` in the cache key. (Source: alternatives item 9)

<details>
<summary>🔮 Prompt for AI Agents</summary>

```
Verify each finding against the current code and only fix it if needed.

In `daydream/training/backfill_cache.py`, `BackfillCache` key does not include `kwargs`, risking cache collisions: The cache key appears to be `(repo, endpoint)` based on test patterns. If two calls share the same endpoint but differ in `kwargs` (e.g., different `method` or `paginate` parameters), the cache returns stale data. Not triggered in current usage since the labeler always calls endpoints with consistent kwargs, but it's a correctness hazard for future callers. Include `json.dumps(kwargs, sort_keys=True)` in the cache key. (Source: alternatives item 9)
```

</details>

</blockquote></details>

</blockquote></details>

<details>
<summary>🔮 Prompt for all review comments with AI agents</summary>

```
Fix the 11 review comment(s) posted on this PR.

If the /beagle-core:fetch-pr-feedback skill is available, run:
/beagle-core:fetch-pr-feedback --pr 95

Otherwise, fetch the comments manually:
1. gh api repos/existential-birds/daydream/pulls/95/comments
2. gh api repos/existential-birds/daydream/issues/95/comments

These endpoints return all comments on the PR. Focus on the most
recent review — ignore older review threads that have already been
addressed. For each comment: read the referenced file, verify the
finding against the current code, and fix it if valid. Skip false
positives. Commit all fixes when done.
```

</details>

<details>
<summary>ℹ️ Review info</summary>

- **Model:** mixed — see breakdown
- **Cost:** $27.98
- **Tokens:** 140 in (140 cached, 100% hit) → 67,106 out
- **Steps / tool calls:** 11 / 204

<details><summary>Per-phase breakdown</summary>

| Phase | Model | Tools | Input (cached) | Output | Cost |
|---|---|---|---|---|---|
| Exploration | mixed | 35 | 15 (100%) | 17,111 | $1.44 |
| Understand Intent | claude-opus-4-6 | 13 | 10 (100%) | 2,853 | $2.89 |
| Alternatives | claude-opus-4-6 | 65 | 33 (100%) | 11,949 | $10.42 |
| Deep Review | claude-opus-4-6 | 62 | 47 (100%) | 17,855 | $12.30 |
| Parse Feedback | claude-haiku-4-5-20251001 | 1 | 12 (100%) | 1,363 | $0.05 |
| Verify Recommendations | claude-sonnet-4-6 | 28 | 23 (100%) | 15,975 | $0.88 |

</details>

<sub>Generated by daydream v0.17.0</sub>

</details>

<sub>🧙 Posted by [daydream v0.17.0](https://github.com/existential-birds/daydream)</sub>

File: daydream/training/export.py [hunk 1]
```diff
@@ -0,0 +1,744 @@
...
+            try:
+                resolved = materialize_base_sha(manifest_path, repo_clone=repo_clone)
+            except OSError:
+                resolved = None
+            if resolved is not None:
+                # Refresh code_ctx from the now-updated manifest on disk.
+                try:
+                    refreshed = json.loads(manifest_path.read_text(encoding="utf-8"))
+                except (OSError, json.JSONDecodeError):
+                    refreshed = None
```
[the remark this answers is not in the document]
Comment (anderskev) [reply daydream/training/export.py:279]: @coderabbitai Fixed in 856efb4. Replaced the four silent `except ... = None` sites in `_build_record` (base_sha materialize, manifest re-read, review-output OSError branches, and rubric_json parse) with `warnings.warn(..., stacklevel=2)` matching the existing `outcome_labels` pattern. `FileNotFoundError` on `review-output.md` stays silent — legitimately-absent file is expected — but other `OSError`s now warn.

File: daydream/training/labeler_signals.py [hunk 2]
```diff
@@ -0,0 +1,383 @@
...
+        file_at_fetcher: Returns file content at a given SHA.
+        window_days: Lookback window for upstream commits.
+
+    Returns:
+        :class:`FixAppliedSignal` with ``verdict="unknown"`` when the
+        commit window is empty, ``"not_applied"`` when no recommended
+        files were touched or fewer than half of the parsed hunks
+        appear post-window, and ``"applied"`` otherwise.
+        ``hunks_applied``, ``hunks_total``, and ``window_commits`` are
+        always populated.
```
[the remark this answers is not in the document]
Comment (anderskev) [reply daydream/training/labeler_signals.py:238]: @coderabbitai Fixed in 856efb4. Added `Raises:` blocks to `fix_applied_signal` and `local_commit_applied_signal` covering `KeyError` (row lookups), `OSError` (`diff.patch` read), and propagated fetcher-callable errors.

File: daydream/training/export.py [hunk 3]
```diff
@@ -0,0 +1,786 @@
...
+    if not code_ctx.get("base_sha") and manifest_path is not None:
+        repo_clone = _resolve_repo_clone(manifest_row.get("repo_slug"))
+        if repo_clone is not None:
+            try:
+                resolved = materialize_base_sha(manifest_path, repo_clone=repo_clone)
+            except OSError as exc:
+                warnings.warn(
+                    f"Session {manifest_row.get('session_id')!r}: base_sha backfill failed for {manifest_path}: {exc}",
+                    stacklevel=2,
+                )
```
[the remark this answers is not in the document]
Comment (anderskev) [reply daydream/training/export.py:268]: @coderabbitai Fixed in 315cabb — broadened the call-site `except` to `(OSError, json.JSONDecodeError, GitError)` and corrected `materialize_base_sha`'s `Raises:` docstring so it accurately lists all three (the prior claim that git failures only return `None` was wrong — `merge_base()` does raise `GitError` on subprocess timeout / missing git binary).
